### PR TITLE
[wip] recording instance ids 

### PR DESCRIPTION
**Pull request by yvsriram**
## Motivation and Context

Keep track of object instances and group different views
image

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

# Checklist:

- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] I ran `tests/hw_manual_test.py` on hardware and observed that robot behavior looks normal.

### CODE REVIEW

File: src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py [hunk 1]
```diff
@@ -590,7 +745,12 @@ def _update_global_map_and_pose_for_env(
         """Update global map and pose and re-center local map and pose for a
         particular environment.
         """
+
         global_map[e, :, lmb[e, 0] : lmb[e, 1], lmb[e, 2] : lmb[e, 3]] = local_map[e]
+        if self.record_instance_ids:
+            global_map = self.add_instance_ids_to_global_map(
```
Comment (theophilegervet) [inline src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py:751]: So clustering the map into object instances doesn't happen at every step but instead when we update the global map. This is not how I imagined it but it makes sense

**Review by theophilegervet — COMMENTED** [review]
Looks clean, well done! Could you add visualization so we can sanity check that it works as expected before we merge?

File: projects/habitat_objectnav/configs/agent/hm3d_eval.yaml [hunk 2]
```diff
@@ -44,6 +44,8 @@ AGENT:
     dilate_obstacles: False
     dilate_size: 3
     dilate_iter: 1
+    record_instance_ids: True
```
Comment (cpaxton) [inline projects/habitat_objectnav/configs/agent/hm3d_eval.yaml:47]: What happens if these new fields aren't filled out? Do we need them in the real robot configs?

File: src/home_robot/home_robot/agent/objectnav_agent/objectnav_agent.py [hunk 3]
```diff
@@ -345,17 +349,28 @@ def _preprocess_obs(self, obs: Observations):
...
+        if self.store_all_categories_in_map:
+            semantic = obs.semantic
+            obj_goal_idx = obs.task_observations["object_goal"]
+            if "start_recep_goal" in obs.task_observations:
+                start_recep_idx = obs.task_observations["start_recep_goal"]
+            elif "end_recep_goal" in obs.task_observations:
+                end_recep_idx = obs.task_observations["end_recep_goal"]
+        else:
+            semantic = np.full_like(obs.semantic, 4)
+            obj_goal_idx, start_recep_idx, end_recep_idx = 1, 2, 3
```
Comment (cpaxton) [inline src/home_robot/home_robot/agent/objectnav_agent/objectnav_agent.py:361]: Should these be constants somewhere?

File: src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py [hunk 4]
```diff
@@ -577,6 +586,152 @@ def _update_local_map_and_pose(
 
         return current_map, current_pose
 
+    def cluster_binary_map(self, binary_map):
```
Comment (cpaxton) [inline src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py:589]: type hints and docstring

File: src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py [hunk 5]
```diff
@@ -577,6 +586,152 @@ def _update_local_map_and_pose(
...
+        # Get the instances from the global map within the local map's region
+        global_instances_within_local = global_instances[x_start:x_end, y_start:y_end]
+
+        # Create a mapping of local instance IDs to global instance IDs
+        instance_mapping = {}
+
+        max_instance_id = torch.max(global_instances)
+
+        # Associate instances in the local map with corresponding instances in the global map
+        for local_instance_id in torch.unique(extended_local_labels):
```
Comment (cpaxton) [inline src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py:663]: This is a very long function that would probably be easier to follow if it were broken up into a couple sub-functions - e.g. one that populates the instance_mapping dict here

File: src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py [hunk 6]
```diff
@@ -577,6 +586,152 @@ def _update_local_map_and_pose(
...
+                # If there is a corresponding instance in the global map, pick the first one and associate it
+                global_instance_id = unique_overlapping_instances[0].item()
+                instance_mapping[local_instance_id.item()] = global_instance_id
+                print("old instance: ", global_instance_id)
+            else:
+                # If there are no corresponding instances, create a new instance
+                new_global_instance_id = max_instance_id + 1
+                instance_mapping[local_instance_id.item()] = new_global_instance_id
+                max_instance_id += 1
+                print("new instance: ", new_global_instance_id)
```
Comment (cpaxton) [inline src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py:689]: hide these behind a "verbose" argument or param?

**Review by cpaxton — CHANGES_REQUESTED** [review]
Some questions + minor feedback

File: projects/habitat_objectnav/configs/agent/hm3d_eval.yaml (hunk 2 above)
Comment (yvsriram) [reply projects/habitat_objectnav/configs/agent/hm3d_eval.yaml:47]: added default values for now and tested

File: src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py (hunk 5 above)
Comment (yvsriram) [reply src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py:711]: Separated

File: src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py [hunk 7]
```diff
@@ -577,6 +586,152 @@ def _update_local_map_and_pose(
...
+        global_instances_in_local = np.vectorize(instance_mapping.get)(
+            local_instances.cpu().numpy()
+        )
+        global_instances[x1:x2, y1:y2] = torch.tensor(
+            global_instances_in_local, dtype=torch.int64, device=global_instances.device
+        )
+        return global_instances
+
+    def add_instance_ids_to_global_map(
+        self, e: int, global_map: Tensor, local_map: Tensor, lmb: Tensor
```
[the remark this answers is not in the document]
Comment (yvsriram) [reply src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py:740]: Added using ChatGPT

File: src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py (hunk 6 above)
Comment (yvsriram) [reply src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py]: Removed debug print statements

File: src/home_robot_sim/home_robot_sim/env/habitat_objectnav_env/visualizer.py [hunk 8]
```diff
@@ -195,6 +195,51 @@ def get_semantic_vis(self, semantic_map, rgb_frame=None):
...
+            )
+
+        return instance_map_combined, instances_per_category
+
+    def replace_semantic_palette_with_instance_palette(
+        self,
+        input_colors: List[int],
+        num_instances: List[int],
+        delta: Optional[int] = 20,
+    ) -> List[int]:
```
Comment (cpaxton) [inline src/home_robot_sim/home_robot_sim/env/habitat_objectnav_env/visualizer.py:229]: please document what is going on here. how would i use this? what is delta doing?

File: src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py [hunk 9]
```diff
@@ -577,6 +586,196 @@ def _update_local_map_and_pose(
...
+    ) -> Tensor:
+        """
+        Adds instance IDs to the global map based on the local map.
+
+        Args:
+            e (int): The index of the environment.
+            global_map (Tensor): The global map tensor.
```
Comment (cpaxton) [inline src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py:747]: this is actually not a great description - assuming global map is the entire world map as a 2d map with channels representing object values?

File: src/home_robot_sim/home_robot_sim/env/habitat_objectnav_env/visualizer.py (hunk 8 above)
Comment (cpaxton) [reply src/home_robot_sim/home_robot_sim/env/habitat_objectnav_env/visualizer.py:229]: delta is actually pretty confusing to me, not immediately obvious even after looking at the code why this is a param

Comment (cpaxton) [reply src/home_robot_sim/home_robot_sim/env/habitat_objectnav_env/visualizer.py:229]: ah i see, delta is a step size through color space... it's a bit odd but i get it

**Review by cpaxton — APPROVED** [review]
lgtm, thanks for refactoring + resolving comments. some documentation could be improved but looks ok to me.

**Comment by cpaxton** [comment]
@keke-220 maybe you could take a look too?

**Comment by cpaxton** [comment]
@keke-220 maybe you can take a look?

**Review by keke-220 — APPROVED** [review]
LGTM. Interested to see the visualization.

File: src/home_robot_sim/home_robot_sim/env/habitat_objectnav_env/visualizer.py (hunk 8 above)
Comment (yvsriram) [reply src/home_robot_sim/home_robot_sim/env/habitat_objectnav_env/visualizer.py]: Removed this function now. The colors were too close.

**Review by theophilegervet — APPROVED** [review]
Overall this PR looks great: good structure and clean code, well done @yvsriram!

A few improvements before we merge:
- Can we move the `instance_tracking_modules.py` file out of `objectnav_agent` into `mapping/semantic` given it's a map and it's not specific to object nav? This is important as it creates circular imports otherwise (object nav agent imports mapping imports object nav agent)
- I raised a PR from a fork with small improvements I made while playing with the code to review it: https://github.com/facebookresearch/home-robot/pull/281. We can merge it if you're happy with the changes
- Can we replace `cat X` with the category name in the visualization? 
- Can we visualize the instance memory at the end of the episode? As a v0, we could add a simple visualization method to your `InstanceMemory` that dumps all of its content to a folder so we can inspect it. We could call this function at the end of the `eval_episode.py` in the `habitat_objectnav` project

**Pull request merged**